src/instrumentation/sdk.go

```diff
@@ -18,497 +18,80 @@ package instrumentation
-	k8sResources[semconv.K8SPodUIDKey] = string(pod.UID)
-	k8sResources[semconv.K8SNodeNameKey] = pod.Spec.NodeName
-	k8sResources[semconv.ServiceInstanceIDKey] = createServiceInstanceId(ns.Name, pod.Name, pod.Spec.Containers[index].Name)
-	i.addParentResourceLabels(ctx, newrelic.Spec.Resource.AddK8sUIDAttributes, ns, pod.ObjectMeta, k8sResources)
-	for k, v := range k8sResources {
-		if !existingRes[string(k)] && v != "" {
-			res[string(k)] = v
-		}
-	}
-	return res
-}
-
-func (i *sdkInjector) addParentResourceLabels(ctx context.Context, uid bool, ns corev1.Namespace, objectMeta metav1.ObjectMeta, resources map[attribute.Key]string) {
-	for _, owner := range objectMeta.OwnerReferences {
-		switch strings.ToLower(owner.Kind) {
-		case "replicaset":
-			resources[semconv.K8SReplicaSetNameKey] = owner.Name
-			if uid {
-				resources[semconv.K8SReplicaSetUIDKey] = string(owner.UID)
-			}
-			// parent of ReplicaSet is e.g. Deployment which we are interested to know
-			rs := appsv1.ReplicaSet{}
-			nsn := types.NamespacedName{Namespace: ns.Name, Name: owner.Name}
-			backOff := wait.Backoff{Duration: 10 * time.Millisecond, Factor: 1.5, Jitter: 0.1, Steps: 20, Cap: 2 * time.Second}
-
-			checkError := func(err error) bool {
-				return apierrors.IsNotFound(err)
-			}
-
-			getReplicaSet := func() error {
-				return i.client.Get(ctx, nsn, &rs)
-			}
-
-			// use a retry loop to get the Deployment. A single call to client.get fails occasionally
-			err := retry.OnError(backOff, checkError, getReplicaSet)
+			injector.ConfigureClient(i.client)
+			injector.ConfigureLogger(i.logger.WithValues("injector", injector.Language()))
+			i.logger.V(1).Info("injecting instrumentation into pod",
+				"agent_language", inst.Spec.Agent.Language,
+				"newrelic-namespace", inst.Namespace,
+				"newrelic-name", inst.Name,
+			)
+			//@todo: add panic handler
+			newPod, err := injector.Inject(ctx, inst, ns, pod)
 			if err != nil {
-				i.logger.Error(err, "failed to get replicaset", "replicaset", nsn.Name, "namespace", nsn.Namespace)
-			}
-			i.addParentResourceLabels(ctx, uid, ns, rs.ObjectMeta, resources)
-		case "deployment":
-			resources[semconv.K8SDeploymentNameKey] = owner.Name
-			if uid {
-				resources[semconv.K8SDeploymentUIDKey] = string(owner.UID)
-			}
-		case "statefulset":
-			resources[semconv.K8SStatefulSetNameKey] = owner.Name
-			if uid {
-				resources[semconv.K8SStatefulSetUIDKey] = string(owner.UID)
-			}
-		case "daemonset":
-			resources[semconv.K8SDaemonSetNameKey] = owner.Name
-			if uid {
-				resources[semconv.K8SDaemonSetUIDKey] = string(owner.UID)
-			}
-		case "job":
-			resources[semconv.K8SJobNameKey] = owner.Name
-			if uid {
-				resources[semconv.K8SJobUIDKey] = string(owner.UID)
-			}
-		case "cronjob":
-			resources[semconv.K8SCronJobNameKey] = owner.Name
-			if uid {
-				resources[semconv.K8SCronJobUIDKey] = string(owner.UID)
+				i.logger.Error(err, "Skipping Java agent injection")
+				continue
```

Comment:
Let's change the logger Error from explicitly calling out Java error

Reply:
For example, It looks like this function is trying to inject `Python` in the second iteration if `Java` was already injected into the `pod`. 

I thought we were not allowing this scenario. We should write some tests to capture this scenario if we are allowing this.

Reply:
fixed